### ISSUE TITLE
List recipes as tree with `just --tree`.

### ISSUE THREAD

**Issue by tducret** [casey/just#1853]
Hi there!
Is there a way to quickly check the recipe dependencies?
I checked the [Listing Available Recipes](https://just.systems/man/en/chapter_24.html) page but it's not mentioned.

If not existing, I think it could be worth to implement.
Let's take an example.
Based on this `Justfile` found in the [documentation](https://just.systems/man/en/chapter_40.html) :
```sh
# Justfile
a:
  echo 'A!'

b: a
  echo 'B start!'
  just c
  echo 'B end!'

c:
  echo 'C!'
```

We could execute the following command :
```sh
just --list --graph
```

And we could clearly see the recipe dependencies :
```sh
├── a
├── b
│   ├── a
│   └── c
└── c
```

What do you think?

**Comment by casey** [comment]
Love this idea! I think calling it `just --tree` would be good, similar to the CLI `tree` command that prints a directory tree.

**Comment by VladimirMarkelov** [comment]
Displaying calls like "just c" as a dependency in the tree can be difficult. What to do in a case:

```
bin := "just" // it can be a conditional here
a:
   echo "a"
    {{bin}} b
b:
  echo "b"
```

or in this case:

```
section := if windows { "build-win" } else { "build-lin" }
a:
   just {{section}}
  echo "Done"
build-win:
  ...
build-lin:
  ...
```

I can think of more complex cases, but even these ones are tricky.

**Comment by casey** [comment]
Oh sorry, I missed that part. Yeah, I think that displaying dependencies which are called inside recipes is not possible.

**Comment by gyreas** [comment]
> Oh sorry, I missed that part. Yeah, I think that displaying dependencies which are called inside recipes is not possible.

Perhaps we adjust the definition of dependency a bit? To me, `lah := "blah"` is an assignment statement, which shouldn't be a dependency, right? Otherwise, this idea has flopped since assignments are (almost) always used in recipes.

Can you consider this point too?

**Comment by jraygauthier** [comment]
That'd be especially nice as a mean to convert of a top level just task sequencing many sub-tasks into various steps of a CI pipeline (assuming a pipeline system that can accept dynamically generated steps) by taking all sub steps (level 1 only). It might even be possible to run those step in parallel :)

That'd make the repository's justfile the single source of truth for what validations to perform on the repository.  

In order to do that however, the output would have to be machine readable, potentially the usual `--json` print option?

**Comment by NomarCub** [comment]
There's another issue that included a visual graph: <https://github.com/casey/just/issues/2033>. People shared some scripts that can do visual graphs from just's output.
Either that or this kind of graph would be really useful!
